src/utp_stream.cpp

```diff
@@ -3090,18 +3102,15 @@ void utp_socket_impl::do_ledbat(const int acked_bytes, const int delay
 		, scaled_gain / double(1 << 16), int(m_cwnd >> 16)
 		, int(m_slow_start));
 
-	// if scaled_gain + m_cwnd <= 0, set m_cwnd to 0
-	if (-scaled_gain >= m_cwnd)
-	{
-		m_cwnd = 0;
-	}
+	// don't drop below 2*MSS. This behavior is from rfc6817 (LEDBAT). This differs
+	// from BEP 29 which allows cwnd to drop to 0, however this way avoids needing
+	// to wait until the next timeout to resume sending.
+	if ((m_cwnd + scaled_gain) >> 16 < 2 * m_mtu)
```

Comment:
`m_cwnd` is `int64_t` fixed point with 16 bit fraction.
`scaled_gain` is also `int64_t` fixed point with 16 bits fraction.
`m_mtu` is `uint16_t` and will be promoted to `int`, which presumably is `int32_t`.
I don't see any risk of overflow here.

However, this does introduce an effective floor of the congestion window. I believe cwnd can still be set to 0 (or < 1) by normal packet loss logic, and the only way to recover is to wait for timeout, and reset cwnd to 1.

However, the odd thing with this logic is that it will immediately snap up cwnd to 2 MSS the first message we receive. Even if `scaled_gain` is positive, and is pushing cwnd *up*, this logic will push it all the way up to 2 MSS unconditionally. I initially got the impression this was meant to be a floor preventing adjusting cwnd *down* too far.

Perhaps it has to work this way, it just seems a bit inconsistent that a timeout would bring cwnd to 1 MSS but this would bring it to 2.

Reply:
I'll rework this into a hard floor for consistency throughout. This would bring the initial `m_cwnd` up to 2*MSS  as well.

LEDBAT clamps to `MIN_CWND` for both latency and loss, and uses `INIT_CWND` to begin with, both of which 'should' be 2*MSS.

Reply:
On second thought, how about only using the floor if cwnd is shrinking. Otherwise, allow it to grow as normal. That way it won't get bumped up to 2*MSS prematurely, though it would if it were to shrink.

A simpler option might be to just use a minimum of 1*MSS here instead of 2, as is currently used elsewhere (on loss/timeout/initial).
	
This is the reasoning behind using a value of 2:

```
   INIT_CWND and MIN_CWND SHOULD both be 2.  An INIT_CWND of 2 should
   help seed FILTER() at the sender when there are no samples at the
   beginning of a flow, and a MIN_CWND of 2 allows FILTER() to use more
   than a single instantaneous delay estimate while not being too
   aggressive.  Slight deviations may be warranted, for example, when
   these values of INIT_CWND and MIN_CWND interact poorly with the
   framing protocol.  However, INIT_CWND and MIN_CWND MUST be no larger
   than the corresponding values specified for TCP [RFC5681].
```

Having a base of 2 for the INIT_CWND makes sense, to get more delay samples at the beginning, however as each uTP packet only contains a single one-way delay sample (LEDBAT seems to allow for combining multiple samples per packet), there may not be a benefit to using 2 instead of 1 for MIN_CWND.